### PR TITLE
Add missing attrs / correct bad attrs on TaskInstancePydantic

### PR DESCRIPTION
**Pull request by dstandish**
One was wrong type; others were missing; others needed a default.

### CODE REVIEW

**Review by potiuk — APPROVED** [review]

**Review by dirrao — CHANGES_REQUESTED** [review]
LGTM. Static checks are failing. Can you look into the static checks?

File: airflow/serialization/pydantic/taskinstance.py [hunk 1]
```diff
@@ -105,12 +105,13 @@ class TaskInstancePydantic(BaseModelPydantic, LoggingMixin):
     trigger_timeout: Optional[datetime]
     next_method: Optional[str]
     next_kwargs: Optional[dict]
-    run_as_user: Optional[str]
+    run_as_user: Optional[str] = None
```
Comment (uranusjr) [inline airflow/serialization/pydantic/taskinstance.py]: Why does this needs a default? (also others added by this PR)

Comment (potiuk) [reply airflow/serialization/pydantic/taskinstance.py]: Indeed - seems not needed.

**Review by eladkal — APPROVED** [review]

File: airflow/serialization/pydantic/taskinstance.py (hunk 1 above)
Comment (dstandish) [reply airflow/serialization/pydantic/taskinstance.py]: the reason is you will see errors like this:
```
[2024-03-17T17:48:27.227+0000] {rpc_api_endpoint.py:135} ERROR - Error executing method: airflow.models.dagrun.DagRun.fetch_task_instance.
[2024-03-17T17:48:27.227+0000] {rpc_api_endpoint.py:136} ERROR - 3 validation errors for TaskInstancePydantic
run_as_user
  Field required [type=missing, input_value=<unprintable TaskInstance object>, input_type=TaskInstance]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
task
  Field required [type=missing, input_value=<unprintable TaskInstance object>, input_type=TaskInstance]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
raw
  Field required [type=missing, input_value=<unprintable TaskInstance object>, input_type=TaskInstance]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/api_internal/endpoints/rpc_api_endpoint.py", line 131, in internal_airflow_api
    output_json = BaseSerialization.serialize(output, use_pydantic_models=True)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/serialization/serialized_objects.py", line 550, in serialize
    mod = _pydantic_model_dump(pyd_mod, var)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.11/site-packages/airflow/serialization/serialized_objects.py", line 546, in _pydantic_model_dump
    return model_cls.model_validate(var).model_dump(mode="json")  # type: ignore[attr-defined]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/airflow/.local/lib/python3.11/site-packages/pydantic/main.py", line 503, in model_validate
    return cls.__pydantic_validator__.validate_python(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 3 validation errors for TaskInstancePydantic
run_as_user
  Field required [type=missing, input_value=<unprintable TaskInstance object>, input_type=TaskInstance]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
task
  Field required [type=missing, input_value=<unprintable TaskInstance object>, input_type=TaskInstance]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
raw
  Field required [type=missing, input_value=<unprintable TaskInstance object>, input_type=TaskInstance]
    For further information visit https://errors.pydantic.dev/2.5/v/missing
    
```

and the reason for _that_, i think, is that these attrs are not set initially on TaskInstance.

that's why for example you see code like this:

https://github.com/apache/airflow/blob/9d898684b26a967242e7f685058d885519b6d383/airflow/models/taskinstance.py#L3661

Should we set these in `init` on TI, instead?

Comment (dstandish) [reply airflow/serialization/pydantic/taskinstance.py]: i guess the reason we haven't done that already is it requires a lot of mypy fixes...

Comment (dstandish) [reply airflow/serialization/pydantic/taskinstance.py]: mmm... it's worse than that... there is some logic that uses `hasattr` which assumes (by virtue of has attr self.task) that `self.task` is not None

**Comment by dstandish** [comment]
ok @uranusjr @potiuk @eladkal ...

In response to the question "why do we need to add defaults to pydantic"....

I have have added a default of None to attr `task` on TI

this isn't really a straightforward change because (1) mypy doesn't like it and, more importantly, (2) there is logic in airflow that looks at `hasattr(ti, "task")` to infer that both `hasattr(ti, "task")` and `ti.task is not None`.  So all of this needs to be updated if we are going to avoid adding a default on the pydantic model.  I have done so in this PR.

can you take a look and let me know what you think is the better approach?  i.e. should we proceed with the changes i've just made, or go back and stick with just adding a default on the pydantic model.

File: airflow/serialization/pydantic/taskinstance.py (hunk 1 above)
Comment (dstandish) [reply airflow/serialization/pydantic/taskinstance.py]: ok i think i mostly resolved this. it's quite ugly (cus so much mypy assertion).  i'm not sure there's really an alternative, apart from giving it a default in pydantic. PTAL. i can split into diff PR if desired.

**Review by potiuk — APPROVED** [review]
LGTM

**Review by vincbeck — APPROVED** [review]

**Pull request merged**